chain/ethereum/src/ingestor.rs

```diff
@@ -188,31 +198,62 @@ impl PollingBlockIngestor {
             .await
             .map(|missing| missing.map(|h256| h256.into()))
             .map_err(|e| {
-                error!(self.logger, "failed to update chain head");
+                error!(logger, "failed to update chain head");
                 IngestorError::Unknown(e)
             })
     }
 
-    async fn latest_block(&self) -> Result<BlockPtr, IngestorError> {
-        self.eth_adapter
-            .latest_block_header(&self.logger)
+    async fn latest_block(
+        &self,
+        logger: &Logger,
+        eth_adapter: &Arc<EthereumAdapter>,
+    ) -> Result<BlockPtr, IngestorError> {
+        eth_adapter
+            .latest_block_header(&logger)
             .compat()
             .await
             .map(|block| block.into())
     }
+
+    async fn eth_adapter(&self) -> anyhow::Result<Arc<EthereumAdapter>> {
+        self.chain_client
+            .rpc()?
+            .cheapest()
+            .await
+            .ok_or_else(|| graph::anyhow::anyhow!("unable to get eth adapter"))
+    }
```

Comment:
do_pool will loop inside IIRC and will only exit if there was an error, So if this will only be called on a retry. Before, if the adapter was broken it would never be replaced. Where are these methods taken an eth_adapter? They are probably called after selection, most place that do adapter selection should have a ChainClient IIRC and then inner methods just get the simpler eth_adapter so they don't need to deal with error. That's the general pattern, this was one of the few places where this behavior hadn't been implemented.

Reply:
I meant methods like `latest_block` or `ingest_block` just preceding this.

Reply:
sorry I'm not sure I follow, once the ingestor fetches the adapter it will re-use the same one until there is an error, in which case we would get a new one which goes through the selection that's already implemented everywhere else. Did I miss some part of it?